### ISSUE TITLE
Support for Conditional (ternary) Operator

### ISSUE THREAD

**Issue by queequac** [silentmatt/expr-eval#45]
Support for the conditional operator would be really great. Currently I added an if-function which accepts the condition and two values... but using `<expression> ? <if-true> : <if-false>` would feel much more natural.

Since this is a ternary operator and the parser can only handle unary and binary operators at the moment, this might be not that easy.

**Comment by silentmatt** [comment]
That's a good idea. I don't think it should be too difficult to add ternary operators.

**Comment by queequac** [comment]
Hi Matt, sounds really good!
Unfortunately, for me it would be difficult to add this. Guess you have a better feeling regarding ternary operators, because you know your library's internals better better than me.

Will you have a look at it, or any hint how to achieve this?

**Comment by silentmatt** [comment]
I started playing with it and it kind of turned into a rabbit hole because of how the parser works. It's basically a hand-written state machine, so changes tend to cascade into more and more changes, with lots of different cases to account for. I've been meaning to replace it with a recursive descent parser anyway, so I started on that, and it's almost ready. It's passing all the current tests, and even some new ones that the old parser failed on, so I'm pretty confident it'll be ready to go soon.

**Comment by queequac** [comment]
Hi Matt. So my gut feeling was not completely wrong... but even better if a new parser could become more powerful and more easy to read. Hopefully the API remains the same or migration is simple.

You could provide a preview on a separate branch, once you are done. I would offer testing the new parser. I am using your library for looots of expressions and performance is king. I am very nosy if the new parser is as performant as the old one - or even faster or slower. So if you are interested in early feedback before publishing an official release, just let me know!

**Comment by silentmatt** [comment]
Yes, your gut feeling was pretty close I think :) In theory, it shouldn't have been difficult, but there was a lot of error-prone state management with the old parser.

The new parser shouldn't affect the external API at all. I did make some changes to the internal representations, that should only cause problems if you're manipulating the expressions yourself. I was hoping to get it out this weekend, but other stuff came up and I didn't quite make it..., I can definitely push up a new branch tonight though.

I haven't compared the performance, but I don't expect it would be too different. I'd definitely appreciate any feedback, once I get it pushed up. I've actually wanted to add some performance tests, so I'll try to work on that too.

**Comment by silentmatt** [comment]
I pushed a new branch called [recursive-descent](https://github.com/silentmatt/js-expression-eval/tree/recursive-descent) with the new parser if you want to try it out.

**Comment by queequac** [comment]
Hi Matt, thanks for the branch, I will give it a try tomorrow.
We are using expressions in very large quantity structures, so it should immediately give an indication whether performance changed into one or the other direction. :)

Unfortuntely we are not using the public API only... so I have to modify the library in a way it works in my project again. (Renaming some math-operations, removing disallowed operators, .. that's just the easy things.) Hopefully this can be done in the new library as easy as before. Have seen operatods are no longer on an `ops `or `ops2 `object but just hardcoded in functions such as `isOperator`.

**Comment by silentmatt** [comment]
I started working on some performance tests and noticed about a 10% decrease in parsing speed on the tests I have so far, but I found a quick optimization that seems to have gotten it back to where it was, and possibly slightly better. That's from my very unscientific tests so far, so your mileage may vary.

Let me know if you have any questions or difficulties with your changes.

**Comment by queequac** [comment]
Hi Matt, I did my necessary changes and it works again... unfortunately the performance decrease is significant.

We have often round about 60 expressions being parsed and then evaluated with varying variables. This ends up in roughly 25k evaluations in total. So I assume the performance loss is in the evaluation (since the upfront parsing can be neglected).

On my machine the old library did this job in 800ms. The new library takes 1400ms. :(
That would kill all NFR tests, so I will keep my fingers from the new implementation at the moment.

Maybe I can dig into the code and try to find out where exactly the time is lost compared to the old one.

**Comment by queequac** [comment]
While digging in your code... realized there are lots of issues back that were solved in the old library in the past.
E.g. using E and PI instead 'E' and 'PI' within the consts prevents any variables starting with E.
That does not have any impact on performance, but some of my tests simply don't run anymore. Argh. ;-)

**Comment by queequac** [comment]
Got the evaluation from 1400ms down to 1100ms... 
But the new parser seems to create more tokens than the old one, maybe that's also part of my performance problem, since more tokens have to be processed on evaluation.

**Comment by queequac** [comment]
Hi Matt, most of the performance gain I got from changing the usage of hasValue/getValue in the evaluate function.
Since the former getValue did not check if the entire chain within objects is really available, I changed it slightly and return undefined if there is no value. See #50 for my changes.

Furthermore I do use the built-in indexOf from the Array object. I know, that one is not available for legacy browsers - but I don't support them anyway. Since this is not significantly faster, I did not add it to my pull-request.
